### ISSUE TITLE
SPIKE: Investigate Caching the Homepage

### ISSUE THREAD

**Issue by rotated8** [emory-libraries/blacklight-catalog#1056]
One avenue to improve homepage performance is to cache the data. With this spike, we would like to know:

- [x] How much time does the current Solr request take?
- [x] Can we cache the entire homepage? What are the pros, cons, and technically tricky bits of doing this?
- [x] Can we cache the Solr response for the homepage? What are the pros, cons, and technically tricky bits of doing this?
- [ ] If we cache the Solr response, should we cache it when the application receives it, or when Solr sends it?
- [x] Passenger is blocking for ~2.5 seconds for each request. What's up with that?

In the end, please make a recommendation for our path forward.

**Comment by bwatson78** [comment]
Q: How much time does the current Solr request take?
A: On Test, which mimics the size of Production, the `Blacklight::SearchService#search_results` process took a total of 17.16 ms.

Q: Can we cache the entire homepage? What are the pros, cons, and technically tricky bits of doing this?
A: Short answer: no. Long answer: there are many caching mechanisms that are available to us, including the built-in `Rails.cache` processing. The problems about caching the homepage are tied into the fact that pulling search results and delivering the homepage are processed from the same controller method: `:index`. When the user visits root/home, the index action senses there are no queries or facets, performs a blank search, blocks displaying of the default list of ten results, and instead delivers the homepage HTML. If we blindly cache the homepage, we'd essentially be caching a massive list of unique queries, which leads me to the next question...

Q: Can we cache the Solr response for the homepage? What are the pros, cons, and technically tricky bits of doing this?
A: Yes, but it could take some work/time to do it right. Just like the fact we have logic that blocks the list when the users' are on the homepage, we'd use that same logic to only cache the empty query response. But this has complications, as well. `Rails.cache` can operate very easily on strings, numbers, or arrays of either, but is finicky when caching class objects. There are Redis gem options, but those involve setting up and debugging a major structural addition to the application.
Yet, there is also the consideration of stale data concerning facets--either missing expected values or facet counts that differ as soon as one visits the next search results page. 
There is one more thing to bring up here: if we move forward with any caching of the Solr response, the flamegraph indicates that it represents 17.16ms out of 2.94 seconds, which is a 0.58% of the total time. It's not going to make a major dent in our load time.

Q: Passenger is blocking for ~2.5 seconds for each request. What's up with that?
A: The process that Passenger is hanging up on is `PhusionPassenger::RequestHandler#wait_until_termination_requested`, which is unique to the server side of the application. The Rails extent of interacting with this software is a Capistrano gem that allows the deployment process to restart the server. If there is any tweaking of settings that we can do, I believe it would fall to Middleware to perform.

As for the question about caching the Solr response before/after the application receives it, I'm not aware of what could cache the data inbetween Rails and Solr, so I would defer to those with some knowledge of that.

**Comment by abelemlih** [comment]
@rotated8  A few conclusions from research Brad and I did concerning the issues discussed in this ticket:

**How much time does the current Solr request take?**

When it comes to how long retrieving results from Solr take, the flamegraph can be misleading. The best way to find out how long a Solr fetch takes is by looking at the logs, and recording the duration returned in the logs for the Solr fetch. For instance, on the Test instance, on average the page load took around 3 seconds and the Solr fetch took around 2.7 seconds. This explains how the Solr fetch slows down the index page load significantly.

**Can we cache the entire homepage? What are the pros, cons, and technically tricky bits of doing this?**

Yes, but that may not be the feasible solution at this time. From my conversation with @rotated8, it is clear that we currently do not intend on adding any new infrastructure like Redis, which would make using `Rails.cache` easier and more consistent.

**Can we cache the Solr response for the homepage? What are the pros, cons, and technically tricky bits of doing this**?

Yes, we can cache the data returned from Solr, and build the `Blacklight::Solr::Response` manually in the `CatalogController`. Per @rotated8's recommendation, we are going to store the Solr data in the database, and set an expiration date for each entry. By caching, we are not actually going to use the `Rails.cache`, but rather a new model for Solr cache entries. Based on my research so far, this is not difficult to implement.

**Passenger is blocking for ~2.5 seconds for each request. What's up with that?**

We figured Passenger is not blocking ~2.5 seconds for each request, the way the flamegraphs present the time taken by tasks like `Blacklight::SearchService#search_results` is misleading. We confirmed from the logs of Test that it is in fact the Solr fetch that takes too long.
